README.md
# Pokepush

[![Build Status](https://travis-ci.org/dustinblackman/pokepush.svg?branch=master)](https://travis-ci.org/dustinblackman/pokepush)
[![Go Report Card](https://goreportcard.com/badge/github.com/dustinblackman/pokepush)](https://goreportcard.com/report/github.com/dustinblackman/pokepush)

Dead simple Pokemon Go push notifications through Pushbullet when a new Pokemon arrives in a specified locations. Supports multiple users subscribed to desired locations.

## Setup

- Download the binaries for your system from [releases](https://github.com/dustinblackman/pokepush/releases).
- Acquire a pushbullet token from your accounts [settings page](https://www.pushbullet.com/#settings).
- Create config.json next to the binary.
- Start.

**Example Config:**
```json
{
  "users": [
    {
      "name": "Jim",
      "pushbullet_key": "PUSHBULLET_KEY",
      "places": ["Home", "Work"]
    }
  ],
  "places": [
    {"name": "Home", "lat": 51.501476, "lon": -0.140634},
    {"name": "Work", "lat": 51.5159394, "lon": -0.1352864}
  ],
  "logs": "pretty"
}
```

- `users` - An array of users containing their `name`, `pushbullet_key`, an array of `places`. Names of places must match the ones in the `places` array.
- `places` - An array of locations with their `name`, `lat`, and `lon`. You can covert an address to latitude/longitude [here](http://www.latlong.net/convert-address-to-lat-long.html)
- `logs` - Sets log type, supports `pretty` and `json`. *Default is `pretty`.*

## Docker

A docker image is also available over at [Docker Hub](https://hub.docker.com/r/dustinblackman/pokepush/).

## Development

After git pulling the repo in to your `$GOPATH`, executing the following to get a development setup running.

```bash
make deps
make dev
```

## TODO
- Handle errors
- Write tests
- Handle scheduled places to only push notifications during specified times of day

## Credit
- [PokeVision](http://pokevision)

## [License](LICENSE)


CHANGELOG.md
# Changelog

<a name="0.0.1" />
### 0.0.1 (July 29th, 2016)

- Initial Release


pokemon.go
package main

var pokemonIDs = map[string]string{
	"1":   "Bulbasaur",
	"2":   "Ivysaur",
	"3":   "Venusaur",
	"4":   "Charmander",
	"5":   "Charmeleon",
	"6":   "Charizard",
	"7":   "Squirtle",
	"8":   "Wartortle",
	"9":   "Blastoise",
	"10":  "Caterpie",
	"11":  "Metapod",
	"12":  "Butterfree",
	"13":  "Weedle",
	"14":  "Kakuna",
	"15":  "Beedrill",
	"16":  "Pidgey",
	"17":  "Pidgeotto",
	"18":  "Pidgeot",
	"19":  "Rattata",
	"20":  "Raticate",
	"21":  "Spearow",
	"22":  "Fearow",
	"23":  "Ekans",
	"24":  "Arbok",
	"25":  "Pikachu",
	"26":  "Raichu",
	"27":  "Sandshrew",
	"28":  "Sandslash",
	"29":  "Nidoran♀",
	"30":  "Nidorina",
	"31":  "Nidoqueen",
	"32":  "Nidoran♂",
	"33":  "Nidorino",
	"34":  "Nidoking",
	"35":  "Clefairy",
	"36":  "Clefable",
	"37":  "Vulpix",
	"38":  "Ninetales",
	"39":  "Jigglypuff",
	"40":  "Wigglytuff",
	"41":  "Zubat",
	"42":  "Golbat",
	"43":  "Oddish",
	"44":  "Gloom",
	"45":  "Vileplume",
	"46":  "Paras",
	"47":  "Parasect",
	"48":  "Venonat",
	"49":  "Venomoth",
	"50":  "Diglett",
	"51":  "Dugtrio",
	"52":  "Meowth",
	"53":  "Persian",
	"54":  "Psyduck",
	"55":  "Golduck",
	"56":  "Mankey",
	"57":  "Primeape",
	"58":  "Growlithe",
	"59":  "Arcanine",
	"60":  "Poliwag",
	"61":  "Poliwhirl",
	"62":  "Poliwrath",
	"63":  "Abra",
	"64":  "Kadabra",
	"65":  "Alakazam",
	"66":  "Machop",
	"67":  "Machoke",
	"68":  "Machamp",
	"69":  "Bellsprout",
	"70":  "Weepinbell",
	"71":  "Victreebel",
	"72":  "Tentacool",
	"73":  "Tentacruel",
	"74":  "Geodude",
	"75":  "Graveler",
	"76":  "Golem",
	"77":  "Ponyta",
	"78":  "Rapidash",
	"79":  "Slowpoke",
	"80":  "Slowbro",
	"81":  "Magnemite",
	"82":  "Magneton",
	"83":  "Farfetch'd",
	"84":  "Doduo",
	"85":  "Dodrio",
	"86":  "Seel",
	"87":  "Dewgong",
	"88":  "Grimer",
	"89":  "Muk",
	"90":  "Shellder",
	"91":  "Cloyster",
	"92":  "Gastly",
	"93":  "Haunter",
	"94":  "Gengar",
	"95":  "Onix",
	"96":  "Drowzee",
	"97":  "Hypno",
	"98":  "Krabby",
	"99":  "Kingler",
	"100": "Voltorb",
	"101": "Electrode",
	"102": "Exeggcute",
	"103": "Exeggutor",
	"104": "Cubone",
	"105": "Marowak",
	"106": "Hitmonlee",
	"107": "Hitmonchan",
	"108": "Lickitung",
	"109": "Koffing",
	"110": "Weezing",
	"111": "Rhyhorn",
	"112": "Rhydon",
	"113": "Chansey",
	"114": "Tangela",
	"115": "Kangaskhan",
	"116": "Horsea",
	"117": "Seadra",
	"118": "Goldeen",
	"119": "Seaking",
	"120": "Staryu",
	"121": "Starmie",
	"122": "Mr. Mime",
	"123": "Scyther",
	"124": "Jynx",
	"125": "Electabuzz",
	"126": "Magmar",
	"127": "Pinsir",
	"128": "Tauros",
	"129": "Magikarp",
	"130": "Gyarados",
	"131": "Lapras",
	"132": "Ditto",
	"133": "Eevee",
	"134": "Vaporeon",
	"135": "Jolteon",
	"136": "Flareon",
	"137": "Porygon",
	"138": "Omanyte",
	"139": "Omastar",
	"140": "Kabuto",
	"141": "Kabutops",
	"142": "Aerodactyl",
	"143": "Snorlax",
	"144": "Articuno",
	"145": "Zapdos",
	"146": "Moltres",
	"147": "Dratini",
	"148": "Dragonair",
	"149": "Dragonite",
	"150": "Mewtwo",
	"151": "Mew",
}


main.go
package main

import (
	"fmt"
	"io/ioutil"
	"strconv"
	"strings"
	"time"

	"github.com/Sirupsen/logrus"
	"github.com/buger/jsonparser"
	"github.com/kellydunn/golang-geo"
	"github.com/parnurzeal/gorequest"
	"github.com/robfig/cron"
	"github.com/x-cray/logrus-prefixed-formatter"
	"github.com/xconstruct/go-pushbullet"
)

// Places contains the name, point, and polygon variables for every location defined in configs.
type Places struct {
	Name    string
	Point   *geo.Point
	Polygon *geo.Polygon
}

// User contains the the Pushbullet key and subscribed locations for a user.
type User struct {
	PushBulletKey string
	Places        []string
}

const (
	pointRange = 0.04 // Kilometers
)

var (
	degrees = []float64{0, 45, 90, 135, 180, 225, 270, 315}
	places  = []*Places{}
	usedIDs = map[string]int64{}
	users   = []User{}
	version = "dev"
)

func stringInSlice(a string, list []string) bool {
	for _, b := range list {
		if b == a {
			return true
		}
	}
	return false
}

// createPlace creates a Places based of name, lat, and lon.
func createPlace(name string, lat, lon float64) *Places {
	point := geo.NewPoint(lat, lon)
	points := []*geo.Point{}

	for _, degree := range degrees {
		points = append(points, point.PointAtDistanceAndBearing(pointRange, degree))
	}
	polygon := geo.NewPolygon(points)
	return &Places{name, point, polygon}
}

// createMessage creates the message for all found pokemon
func createMessage(pokemon []string) string {
	l := len(pokemon)
	if l == 1 {
		return pokemon[0]
	}

	wordsForSentence := make([]string, l)
	copy(wordsForSentence, pokemon)
	wordsForSentence[l-1] = "and " + wordsForSentence[l-1]
	return strings.Join(wordsForSentence, ", ")
}

// sendToDevices interates over every user to verify if they are subscribed to a location that's detected a pokemon, and then sends a message.
func sendToDevices(place string, pokemon []string) {
	for _, user := range users {
		if stringInSlice(place, user.Places) {
			pb := pushbullet.New(user.PushBulletKey)
			devices, _ := pb.Devices()
			for _, device := range devices {
				pb.PushNote(device.Iden, "Pokemon detected at "+place, createMessage(pokemon))
			}
		}
	}
}

// scanPokevision hits the pokevision API for all subscribed locations, and calls sendToDevices when found.
// TODO: Error handling
func scanPokevision(place *Places) []string {
	lat := strconv.FormatFloat(place.Point.Lat(), 'f', -1, 64)
	lon := strconv.FormatFloat(place.Point.Lng(), 'f', -1, 64)

	_, body, _ := gorequest.New().Get(fmt.Sprintf("https://pokevision.com/map/scan/%s/%s", lat, lon)).End()
	jobID, _ := jsonparser.GetString([]byte(body), "jobId")
	time.Sleep(5000 * time.Millisecond)
	// TODO Handle jobs in progress
	_, body, _ = gorequest.New().Get(fmt.Sprintf("https://pokevision.com/map/data/%s/%s/%s", lat, lon, jobID)).End()

	available := []string{}

	pokemon, _, _, _ := jsonparser.Get([]byte(body), "pokemon")
	jsonparser.ArrayEach(pokemon, func(value []byte, dataType jsonparser.ValueType, offset int, err error) {
		pokemonIDInt, _ := jsonparser.GetInt(value, "pokemonId")
		pokemonID := strconv.FormatInt(pokemonIDInt, 10)
		id, _ := jsonparser.GetString(value, "id")
		expirationTime, _ := jsonparser.GetInt(value, "expiration_time")
		lat, _ := jsonparser.GetFloat(value, "latitude")
		lon, _ := jsonparser.GetFloat(value, "longitude")
		point := geo.NewPoint(lat, lon)

		if _, ok := usedIDs[id]; !ok && place.Polygon.Contains(point) {
			available = append(available, pokemonIDs[pokemonID])
			usedIDs[id] = expirationTime
		}
	})

	fmt.Println(available)

	uniqMap := map[string]bool{}
	uniqAvailable := []string{}
	for _, name := range available {
		uniqMap[name] = true
	}
	for key := range uniqMap {
		uniqAvailable = append(uniqAvailable, key)
	}

	return uniqAvailable
}

// searchPlaces checks all subscribed locations.
func searchPlaces() {
	logrus.Info("Searching places")
	for _, place := range places {
		availablePokemon := scanPokevision(place)
		if len(availablePokemon) >= 1 {
			logrus.Info("Found pokemon at " + place.Name)
			logrus.Info(availablePokemon)
			sendToDevices(place.Name, availablePokemon)
		}
	}
}

// cleanExpiredIDs clears the expired pokemon IDs cache
func cleanExpiredIDs() {
	for id, expireTime := range usedIDs {
		if time.Unix(expireTime, 0).Before(time.Now()) {
			delete(usedIDs, id)
		}
	}
}

// loadConfig loads the config.json file.
// TODO: Error handling
func loadConfig() {
	config, _ := ioutil.ReadFile("./config.json")
	logrus.SetFormatter(new(prefixed.TextFormatter))

	logtype, _ := jsonparser.GetString(config, "log")
	if logtype == "json" {
		logrus.SetFormatter(&logrus.JSONFormatter{})
	}

	logrus.Info("Pokepush v" + version)

	placesJSON, _, _, _ := jsonparser.Get(config, "places")
	jsonparser.ArrayEach(placesJSON, func(value []byte, dataType jsonparser.ValueType, offset int, err error) {
		name, _ := jsonparser.GetString(value, "name")
		lat, _ := jsonparser.GetFloat(value, "lat")
		lon, _ := jsonparser.GetFloat(value, "lon")
		places = append(places, createPlace(name, lat, lon))
	})

	usersJSON, _, _, _ := jsonparser.Get(config, "users")
	jsonparser.ArrayEach(usersJSON, func(value []byte, dataType jsonparser.ValueType, offset int, err error) {
		pushbulletKey, _ := jsonparser.GetString(value, "pushbullet_key")
		userPlacesJSON, _, _, _ := jsonparser.Get(value, "places")

		userPlaces := []string{}
		jsonparser.ArrayEach(userPlacesJSON, func(value []byte, dataType jsonparser.ValueType, offset int, err error) {
			place, _ := jsonparser.ParseString(value)
			userPlaces = append(userPlaces, place)
		})

		users = append(users, User{pushbulletKey, userPlaces})
	})
}

func main() {
	loadConfig()
	c := cron.New()
	c.AddFunc("@every 2m", searchPlaces)
	c.AddFunc("@every 10m", cleanExpiredIDs)
	searchPlaces()

	logrus.Info("Starting cron")
	c.Start()

	select {} // Block forever.
}
